src/cargo/core/compiler/compilation.rs

```diff
@@ -274,7 +274,12 @@ impl<'cfg> Compilation<'cfg> {
         is_rustc_tool: bool,
     ) -> CargoResult<ProcessBuilder> {
         let mut search_path = Vec::new();
+        // this divides our path based on CompileKind: host vs target.
         if is_rustc_tool {
+            search_path.extend(super::filter_dynamic_search_path(
```

Comment:
I believe you've read this comment <https://github.com/rust-lang/cargo/pull/10469#issuecomment-1079503079> and [this zulip topic](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/dynamic.20linkers.20.26.20mixed.20targets). 

If I understand correctly, the issue is like

1. `rustdoc --test` does two things: compiling the example and running it
2. a proc-macro dep uses a shared lib “foo.so” from system
3. A build script builds a shared lib “foo.so” from vendored sources
4. Name collision happens as proc-macro is supposed to use "foo.so" from system not vendored one.

This should be unlikely as ehuss stated out:

> In general, using shared libraries with proc-macros has some bugs, so hopefully people don't actually use them too often. For example, I just discovered that `cargo doc` does not set the search path for build script OUT directories, so any proc-macro that tries to load a shared library generated by a build script will fail.

That being said, we can propably have patch like this PR and document the truth that rustc/cargo don't support proc-macro loading shared libs at this moment.

Reply:
I'll defer to the team. I'm just carrying the change forward from #10469.